File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int x, time;
        public Node(int x, int time) {
            this.x = x;
            this.time = time;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        Queue<Node> que = new LinkedList<>();
        boolean[] visited = new boolean[100001];
        que.add(new Node(N, 0));
        visited[N] = true;
        while(!que.isEmpty()) {
            Node node = que.poll();
            if(node.x == K) {
                System.out.println(node.time);
                return;
            }
            if(node.x * 2 <= 100000 && !visited[node.x * 2]) {
                que.add(new Node(node.x * 2, node.time + 1));
                visited[node.x * 2] = true;
            } 
            if(node.x - 1 >= 0 && !visited[node.x - 1]) {
                que.add(new Node(node.x - 1, node.time + 1));
                visited[node.x - 1] = true;
            } 
            if(node.x + 1 <= 100000 && !visited[node.x + 1]) {
                que.add(new Node(node.x + 1, node.time + 1));
                visited[node.x + 1] = true;
            }
        }
    }
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.arr.length === this.head;
  }
}

const [N, K] = require('fs').readFileSync('/dev/stdin').toString().trim().split(' ').map(Number);
const que = new Queue();
const visited = new Array(100001).fill(false);
que.add({x : N, time : 0});
visited[N] = true;
while(!que.isEmpty()) {
  const node = que.poll();
  if(node.x === K) {
    console.log(node.time);
    return;
  }
  if(node.x * 2 <= 100000 && !visited[node.x * 2]) {
    que.add({x : node.x * 2, time : node.time + 1});
    visited[node.x * 2] = true;
  }
  if(node.x - 1 >= 0 && !visited[node.x - 1]) {
    que.add({x : node.x - 1, time : node.time + 1});
    visited[node.x - 1] = true;
  }
  if(node.x + 1 <= 100000 && !visited[node.x + 1]) {
    que.add({x : node.x + 1, time : node.time + 1});
    visited[node.x + 1] = true;
  }
}